src/algorithms/peter_norving.py

```diff
@@ -0,0 +1,165 @@
+class PeterNorvig:
+
+    def __init__(self):
+        self.digits = '123456789'
+        self.rows = 'ABCDEFGHI'
+        self.colums = self.digits
+        self.squares = self.cross(self.rows, self.colums)
+        self.unitlist = [self.cross(self.rows, colum) for colum in
+                         self.colums] + [self.cross(row, self.colums)
+                for row in self.rows] + [self.cross(row_square,
+                colum_square) for row_square in ('ABC', 'DEF', 'GHI')
+                for colum_square in ('123', '456', '789')]
+        self.units = dict((square, [unit for unit in self.unitlist
+                          if square in unit]) for square in
+                          self.squares)
+        self.peers = dict((square, set(sum(self.units[square], []))
+                          - set([square])) for square in self.squares)
+
+    def cross(self, row_key_list, column_key_list):
+        """
+        Cross product of elements in row_key_list and elements in column_key_list.
+
+        Keyword arguments:
+        row_key_list -- Mostly this will be the row values. i.e  1, 2, 3,etc
+        column_key_list -- Mostly this will be the column values. i.e A, B ,C, etc
+        returns list with the filled keys crossed that will be used as
+        coordinates, i.e.  [1A, 2A ].
+        """
+
+        return [key + value for key in row_key_list for value in column_key_list]
+
+    def parse_grid(self, grid):
+        """
+        Converts grid game string into a dictionary of possible values,
+        {square: digits}, or return False if a contradiction is detected.
+
+        Keyword arguments:
+        grid -- String that contains the sudoku game, i.e:
+                3000800000007000051000000000000003600020 .......
+        returns a dictionary with all the possible values per square. i.e:
+               {'I6': '1379', 'H9': '679', 'I2': '12369' ........
+        """
+
+        values = dict((square, self.digits) for square in self.squares)
+        for (square, digit) in list(self.create_game_dict(grid).items()):
+            if digit in self.digits and not self.assign(values, square,
+                    digit):
+                return False
+        return values
+
+    def create_game_dict(self, grid):
+        """
+        Convert grid game string into a dictionary  of {square: char}
+        with '0' for empties.
+
+        Keyword arguments:
+        grid -- String that contains the sudoku game, i.e:
+                3000800000007000051000000000000003600020 .......
+        returns grid dict with the filled values ready to play.
```

Comment:
returns grid dict? this means that returns a grid different that the one received as parameter? if so what is the expected return value for this? please add one example